### PR TITLE
PRC-156 : Build the PREPROD environment

### PR DESCRIPTION
**Pull request by prabash-moj**
PRC-156 : Build the PREPROD environment

### CODE REVIEW

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/01-rbac.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  name: hmpps-contacts-preprod-admin
+  namespace: hmpps-contacts-preprod
+subjects:
+  - kind: Group
+    name: "github:dps-tech"
+    apiGroup: rbac.authorization.k8s.io
+  - kind: Group
+    name: "github:hmpps-move-and-improve"
+    apiGroup: rbac.authorization.k8s.io
+roleRef:
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/01-rbac.yaml:13]: The service reliability team (SRE) generally want to be added to RBAC for CP please 
```
- kind: Group
    name: "github:hmpps-sre"
    apiGroup: rbac.authorization.k8s.io
````

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/irsa.tf [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# Add the names of the SQS queues & SNS topics which the app needs permissions to access.
+# The value of each item should be the namespace where the queue or topic was created.
+# This information is used to collect the IAM policies which are used by the IRSA module.
+locals {
+  # The names of the SNS topics used and the namespace which created them
+  sns_topics = {
+    "cloud-platform-Digital-Prison-Services-e29fb030a51b3576dd645aa5e460e573" = "hmpps-domain-events-preprod"
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/irsa.tf:7]: This topic ID is the topic in DEV - in preprod this should be
`cloud-platform-Digital-Prison-Services-15b2b4a6af7714848baeaf5f41c85fcd`

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/irsa.tf [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/irsa.tf:9]: I think we're missing the audit queue definitions here? `sqs_queues = {
    "Digital-Prison-Services-preprod-hmpps_audit_queue" = "hmpps-audit-preprod",
  }`

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/irsa.tf [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  sns_policies = { for item in data.aws_ssm_parameter.irsa_policy_arns_sns : item.name => item.value }
+}
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/irsa.tf:11]: `sqs_policies = { for item in data.aws_ssm_parameter.irsa_policy_arns_sqs : item.name => item.value }` ?

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/irsa.tf [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  namespace              = var.namespace # this is also used to attach your service account to your namespace
+  environment_name       = var.environment
+  infrastructure_support = var.infrastructure_support
+}
+
+data "aws_ssm_parameter" "irsa_policy_arns_sns" {
+  for_each = local.sns_topics
+  name     = "/${each.value}/sns/${each.key}/irsa-policy-arn"
+}
+
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/irsa.tf:37]:
```
data "aws_ssm_parameter" "irsa_policy_arns_sqs" {
  for_each = local.sqs_queues
  name     = "/${each.value}/sqs/${each.key}/irsa-policy-arn"
}
```

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/rds-postgresql.tf [hunk 6]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Make sure that you use the latest version of the module by changing the
+ * `ref=` value in the `source` attribute to the latest version listed on the
+ * releases page of this repository.
+ *
+ */
+module "rds" {
+  source = "github.com/ministryofjustice/cloud-platform-terraform-rds-instance?ref=7.0.0"
+
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/rds-postgresql.tf:9]: v7.3.0?  Later versions available.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/rds-postgresql.tf [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  # VPC configuration
+  vpc_name = var.vpc_name
+
+  # RDS configuration
+  allow_minor_version_upgrade  = true
+  allow_major_version_upgrade  = false
+  performance_insights_enabled = false
+  db_max_allocated_storage     = "500"
+  enable_rds_auto_start_stop   = true # Uncomment to turn off your database overnight between 10PM and 6AM UTC / 11PM and 7AM BST.
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/rds-postgresql.tf:18]: max storage probably wants to be higher .. but would need some sizing task to figure out what to set it to.

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/rds-postgresql.tf [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  # db_password_rotated_date     = "2023-04-17" # Uncomment to rotate your database password.
+
+  # PostgreSQL specifics
+  db_engine         = "postgres"
+  db_engine_version = "16"
+  rds_family        = "postgres16"
+  db_instance_class = "db.t4g.micro"
+
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/rds-postgresql.tf:26]: instance class will need to be bigger - but sizing task required for this, so maybe start with `db.t4g.small` ?

File: namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/rds-postgresql.tf [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    name      = "rds-postgresql-instance-output"
+    namespace = var.namespace
+  }
+
+  data = {
+    database_name = module.rds.database_name
+    db_identifier = module.rds.db_identifier
+  }
+}
+
```
Comment (timharrison-moj) [inline namespaces/live.cloud-platform.service.justice.gov.uk/hmpps-contacts-preprod/resources/rds-postgresql.tf:63]: When we build the PROD namespace, the PREPROD namespace needs to export a secret for its RDS credentials too - so not part of this PR, but a reminder to come back and add it here.. see BVLS namespace config for details of how..

**Review by timharrison-moj — COMMENTED** [review]
A few comments added.

**Review by timharrison-moj — COMMENTED** [review]
LGTM - so ready to be posted in the ask-cloud-platform channel - though I think they are auto-reviewed now.

**Comment by prabash-moj** [comment]
Please note I have addressed the requested change by the github-actions[bot] , But it is still showing its pending, can you please sort this for me . thanks. @kyphutruong
